flowapi/flowapi/api_spec.py

```diff
@@ -0,0 +1,248 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import yaml
+from quart import Blueprint, request, jsonify, render_template, current_app
+from zmq.asyncio import Socket
+from flowapi import __version__
+
+blueprint = Blueprint("spec", __name__)
+
+
+def remove_discriminators(spec: dict) -> dict:
+    """
+    Remove any discriminator keys from an openapi spec dict.
+
+    Parameters
+    ----------
+    spec : dict
+        Dict version of an openapi spec
+
+    Returns
+    -------
+    dict
+        The same spec, but with all discriminator keys removed.
+    """
+    newdict = {}
+    for k, v in spec.items():
+        if k != "discriminator":
+            newdict[k] = remove_discriminators(v) if isinstance(v, dict) else v
+    return newdict
+
+
+async def get_spec(socket: Socket, request_id: str) -> dict:
+    """
+    Construct open api spec by interrogating FlowMachine.
+
+    Parameters
+    ----------
+    socket : Socket
+    request_id : str
+        Unique id of the request
+
+    Returns
+    -------
+
+    """
+    msg = {"request_id": request_id, "action": "get_query_schemas"}
+    socket.send_json(msg)
+    #  Get the reply.
+    reply = await socket.recv_json()
+    flowmachine_query_schemas = reply["payload"]["query_schemas"]
+    # Need to mark query_kind as a required field
+    # this is a workaround because the marshmallow-oneOf plugin strips
+    # the query_kind off, which means it can't be required from the marshmallow
+    # side without raising an error
+    for schema, schema_dict in flowmachine_query_schemas.items():
+        try:
+            schema_dict["required"].append("query_kind")
+        except KeyError:
+            pass  # Doesn't have any properties
+    schema = {
+        "openapi": "3.0.1",
+        "info": {
+            "description": "FlowKit Analytical API",
+            "version": __version__,
+            "title": "FlowAPI",
+            "contact": {"email": "flowkit@flowminder.org"},
+            "license": {
+                "name": "MPLv2",
+                "url": "https://www.mozilla.org/en-US/MPL/2.0/",
+            },
+        },
+        "paths": {
+            "/run_query": {
```

Comment:
I think the route should be `"/run"` instead of `"/run_query"`.

How about leveraging `apispec` to directly add the paths to the spec via the dostring? (See [example](https://apispec.readthedocs.io/en/stable/) in their docs.) This would help to avoid mismatches in the route names, and also document things like status codes in the docstring for the route.

Reply:
We'd need to roll something ourselves there - the apispec flask plugin does not, at this time, support blueprints. Is mostly a wrapper around `yaml_utils.load_yaml_from_docstring` tho, so I'll give it a look.

Reply:
🎻